op-service/ctxinterrupt/waiter.go

```diff
@@ -0,0 +1,39 @@
+package ctxinterrupt
+
+import (
+	"context"
+	"fmt"
+)
+
+// waiter describes a value that can wait for interrupts and context cancellation at the same time.
+type waiter interface {
+	waitForInterrupt(ctx context.Context) waitResult
+}
+
+// Waits for an interrupt or context cancellation. ctxErr should be the context.Cause of ctx when it
+// is done. interrupt is only inspected if ctxErr is nil, and is not required to be set.
+type WaiterFunc func(ctx context.Context) (interrupt, ctxErr error)
```

Comment:
I don't like the double-error checking as much; errors are an interface and can be inspected to inherit a certain type of error if needed. But if you believe it's a better approach then I'm willing to go with it.

Reply:
Yeah I think it's better to split possible values where it makes sense and you don't need to conform to an outside interface. Go's catch-all interfaces are great for tunnel values through without requiring changes but risky. I think in this case I separate interrupt/context failure and then combine them again when I need to fit into just an `error`.